python/pyarrow/parquet.py

```diff
@@ -591,23 +592,32 @@ class ParquetManifest(object):
 
     """
     def __init__(self, dirpath, filesystem=None, pathsep='/',
-                 partition_scheme='hive'):
+                 partition_scheme='hive', thread_pool=None):
         self.filesystem = filesystem or _get_fs_from_path(dirpath)
         self.pathsep = pathsep
         self.dirpath = dirpath
         self.partition_scheme = partition_scheme
         self.partitions = ParquetPartitions()
         self.pieces = []
+        self._thread_pool = thread_pool or futures.ThreadPoolExecutor()
```

Comment:
Apparently the default number of threads is ncpu * 5 (https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor). 

I think this needs to be exposed and configurable at the level of `ParquetDataset`, which is how most people will be consuming this. How about adding a parameter `metadata_nthreads` to ParquetDataset and passing that down here? If the user does not pass anything into `ParquetDataset`, we might either want to make `max_workers=1` or a more conservative number of threads. @pitrou do you have an opinion on the default? In the case that we have a fast filesystem, it shouldn't matter much. For slow filesystems and ones with distributed metadata services (like HDFS) this can definitely make things a lot faster

Reply:
I would say keep `max_workers=1` as a default. The optimal number in a given situation will be filesystem- and system-dependent.